src/views/Visualizer/View/Widgets/Widgets.js

```diff
@@ -25,51 +25,55 @@ const getComponent = component => {
   }
 };
 
-export class Widgets extends React.Component {
-  state = {
-    widgets: [],
-  };
+const Widgets = ({
+  widgets,
+  visible,
+  translate,
+  layersTreeState,
+  ...rest
+}) => {
+  const displayedLayers = React.useMemo(() =>
+    (Array.from(layersTreeState, ([k1, k2]) => ({ ...k1, ...k2 }))
+      .filter(({ active }) => active) || []),
+  [layersTreeState]);
 
-  static getDerivedStateFromProps ({ widgets }) {
-    /**
-     * Keep a backup of widgets to have a beautiful animation with content
-     */
-    if (!widgets.length) return null;
-
-    return { widgets };
+  if (!widgets.length) {
+    return null;
   }
 
-  render () {
-    const { visible, translate, ...props } = this.props;
-    const { widgets } = this.state;
-
-    return (
-      <div
-        className={classnames({
-          'widgets-panel': true,
-          'widgets-panel--visible': visible,
+  return (
+    <div
+      className={classnames({
+        'widgets-panel': true,
+        'widgets-panel--visible': visible,
+      })}
+    >
+      <div className="widgets-panel__container">
+        {(widgets).map(({ widget, filters, layer, form, layerLabel }, index) => {
+          const { component } = widget;
+          const displayedLayer = displayedLayers.find(({ label }) => label === layerLabel);
+          const Component = getComponent(component);
```

Comment:
Maybe you can up this line one step and early return null if the result is falsy.

Reply:
I move this part in its own component, what do you thin ?